Add private key Ingress annotations to set private key properties for Certificate

oGi4i opened the pull request:
### Pull Request Motivation

Closes #4997

### Kind

feature

### Release Note

```release-note
Adds private key Ingress annotations to set private key properties for Certificate
```

oGi4i commented:
/kind feature

irbekrm commented:
/asign
/ok-to-test

oGi4i commented:
/retest pull-cert-manager-e2e-v1-24

oGi4i commented:
/test pull-cert-manager-e2e-v1-24

irbekrm reviewed (changes requested):
Thank you very much for working on this @oGi4i !

I've added a couple nits.

Also, something I forgot to check for the previous PR- we should also update [`certNeedsUpdate`](https://github.com/cert-manager/cert-manager/blob/master/pkg/controller/certificate-shim/sync.go#L472-L512) to ensure that the `Certificate` gets updated if the annotation values have changed for private key block of the revision history.

irbekrm commented:
This is a flake 

/test pull-cert-manager-e2e-v1-24

oGi4i commented on pkg/controller/certificate-shim/helper.go:
```diff
@@ -110,5 +110,84 @@ func translateAnnotations(crt *cmapi.Certificate, ingLikeAnnotations map[string]
+
+		switch algorithm {
+		case cmapi.RSAKeyAlgorithm:
+			if size < 2048 {
```
Yeah, I intentionally made the validation stricter because I think of the annotations as an override mechanism.
So I would want to get an error if I override with a value that is actually ignored.

oGi4i replied:
> We also want to error out if the size > 8192 (see [webhook validation for `certificate.spec.privateKey`](https://github.com/cert-manager/cert-manager/blob/master/internal/apis/certmanager/validation/certificate.go#L115)

Thanks, fixed it

oGi4i commented:
> Thank you very much for working on this @oGi4i !
> 
> I've added a couple nits.
> 
> Also, something I forgot to check for the previous PR- we should also update [`certNeedsUpdate`](https://github.com/cert-manager/cert-manager/blob/master/pkg/controller/certificate-shim/sync.go#L472-L512) to ensure that the `Certificate` gets updated if the annotation values have changed for private key block of the revision history.

Thanks for pointing this out. Added new checks to `certNeedsUpdate`

oGi4i commented:
/retest

oGi4i commented:
/retest

irbekrm commented on pkg/controller/certificate-shim/sync.go:
```diff
@@ -508,6 +508,106 @@ func certNeedsUpdate(a, b *cmapi.Certificate) bool {
+		return true
+	}
+
+	aAlgorithm, bAlgorithm := cmapi.RSAKeyAlgorithm, cmapi.RSAKeyAlgorithm
```
Suggestion: do something like `var aAlgorithm, bAlgorithm campi.PrivateKeyAlgorithm` instead.

I understand that `RSA` is the default value so this check is logically correct, but is harder to read as the reader needs to follow that nil value in one cert is not different from RSA value in another etc.
Also this would follow the same approach as other checks in this function i.e [this](https://github.com/cert-manager/cert-manager/blob/master/pkg/controller/certificate-shim/sync.go#L507) where we compare values as they are in the certs, not taking into account defaulting

oGi4i replied:
I actually tried this approach, but I found that this adds additional checks like `if aAlgorithm == "" && bAlgorithm != cmapi.RSAKeyAlgorithm` => `return true`.
Because we have to check the case when one private key does not have a value set (so it effectively has RSA as the default) and the other has RSA set implicitly.

oGi4i replied:
Fixed it

irbekrm commented on pkg/controller/certificate-shim/sync.go:
```diff
@@ -508,6 +508,106 @@ func certNeedsUpdate(a, b *cmapi.Certificate) bool {
+		return true
+	}
+
+	aEncoding, bEncoding := cmapi.PKCS1, cmapi.PKCS1
```
Suggestion: (same as above with the same reasoning) `var aEncoding, bEncoding cmapi.PrivateKeyEncoding`

irbekrm commented on pkg/controller/certificate-shim/sync.go:
```diff
@@ -508,6 +508,106 @@ func certNeedsUpdate(a, b *cmapi.Certificate) bool {
+		bSize = b.Spec.PrivateKey.Size
+	}
+
+	switch aAlgorithm {
```
Suggestion: at this point I think we could just do `if aSize != bSize return false`? I think at this point we should just update the cert spec to be correct- it will be up to actual certificate controllers to determine if the updated spec actually means that there is a need for re-issuance.
Let me know if you agree

oGi4i replied:
Yeah, I agree that here we just want to update the cert spec to be up to date.
I've made all the checks simpler.

irbekrm reviewed (changes requested):
Hi @oGi4i thanks for the changes and all the new test cases!

I've added a couple more suggestions, let me know if that makes sense!

oGi4i commented on pkg/controller/certificate-shim/sync.go:
```diff
@@ -508,6 +508,106 @@ func certNeedsUpdate(a, b *cmapi.Certificate) bool {
+		return true
+	}
+
+	aRotationPolicy, bRotationPolicy := cmapi.RotationPolicyNever, cmapi.RotationPolicyNever
```
Fixed it

irbekrm reviewed (approved):
Thanks a lot for this contribution @oGi4i and for making all the changes!

I think this will be very useful for many users.

/lgtm

irbekrm commented:
It would be good to understand the underlying issue why sometimes a restart is needed.

irbekrm commented:
/test pull-cert-manager-e2e-v1-24
